Fix(node): altda failover to ethda should keep finalizing l2 chain

samlaf opened the pull request:
<!--
Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md
-->

**Description**

Wenwei from Polymer raised this bug with us.

We are currently in the process of adding failover to altda (see [this](https://github.com/ethereum-optimism/specs/issues/434); batcher PR coming soon), and realized that when failover happens, the L2 finalized head completely stalls.

This PR consists of 2 commits:
1. [bb6be8d](https://github.com/ethereum-optimism/optimism/pull/12845/commits/bb6be8dab14ffc09d211cd14a45e2d4de2fba00a): test to show the behavior (fails on this commit)
2. [672f820](https://github.com/ethereum-optimism/optimism/pull/12845/commits/672f8206a95ffd15eabd4bc0663e25b595c1211b): fix damgr so that finalizedHead can keep advancing even after failover

<!--
A clear and concise description of the features you're adding in this pull request.
-->

**Tests**

<!--
Please describe any tests you've added. If you've added no tests, or left important behavior untested, please explain why not.
-->

First commit (see above) contains test that shows the buggy behavior. To run:
```
git checkout bb6be8d
go test -run ^TestAltDA_FinalizationAfterEthDAFailover$ github.com/ethereum-optimism/optimism/op-e2e/actions/altda
```

**Fix**

The problem with the damgr as it is is that
```
d.finalizedHead = d.state.lastPrunedCommitment
```
is always run no matter what. This means that finalization is always driven by altda commitments, but after failover, damgr stops seeing the commitments (because they arent altda commitments anymore). The fix in the second commit is to let the finalization be driven by L1 finalization when there are no altda commitments managed by the damgr.

NOTE: I don't fully understand the subtleties of the damgr and derivation pipeline interactions, so might be doing something wrong here... but seems sound to me, especially since it passes the test.

**Additional Context**
The test is a bit ugly currently (see TODO comment below). Someone with better understanding of op-node and event processing order would surely be able to help me clean this up. I use this patch for debugging event processing, in case this is of use to anyone:
```
diff --git a/op-e2e/actions/altda/altda_test.go b/op-e2e/actions/altda/altda_test.go
index 9d60f2fee..c6ef361d7 100644
--- a/op-e2e/actions/altda/altda_test.go
+++ b/op-e2e/actions/altda/altda_test.go
@@ -194,7 +194,7 @@ func (a *L2AltDA) ActNewL2Tx(t helpers.Testing) {
 //
 // 17 makes sense because challengeWindow=16 and we create 1 extra block before that,
 // and 204 L2blocks = 17 L1blocks * 12 L2blocks/L1block (L1blocktime=12s, L2blocktime=1s)
-func (a *L2AltDA) ActNewL2TxFinalized(t helpers.Testing) {
+func (a *L2AltDA) ActNewL2TxFinalized(t helpers.Testing, logEvents ...bool) {
 	// Include a new l2 batcher transaction, submitting an input commitment to the l1.
 	a.ActNewL2Tx(t)
 	// Create ChallengeWindow empty blocks so the above batcher blocks can finalize (can't be challenged anymore)
@@ -203,7 +203,20 @@ func (a *L2AltDA) ActNewL2TxFinalized(t helpers.Testing) {
 	// TODO: understand why we need to drain the pipeline before AND after actL1Finalized
 	a.sequencer.ActL2PipelineFull(t)
 	a.ActL1Finalized(t)
-	a.sequencer.ActL2PipelineFull(t)
+	if logEvents == nil {
+		a.sequencer.ActL2PipelineFull(t)
+	} else {
+		// Log all events until the end of the pipeline
+		count := 0
+		a.sequencer.ActL2EventsUntil(t, func(ev event.Event) bool {
+			count++
+			a.log.Info("new event", "event", ev, "count", count)
+			if count == 100 {
+				return true
+			}
+			return false
+		}, 100, false)
+	}
 
 	// Uncomment the below code to observe the behavior described in the TODO above
 	// syncStatus := a.sequencer.SyncStatus()
@@ -680,5 +693,12 @@ func TestAltDA_FinalizationAfterEthDAFailover(gt *testing.T) {
 		ssAfter := harness.sequencer.SyncStatus()
 		// Even after failover, the finalized head should continue advancing normally
 		require.Equal(t, ssBefore.FinalizedL2.Number+diffL2Blocks, ssAfter.FinalizedL2.Number)
+
+		harness.log.Info("Sync status before",
+			"unsafeL1", ssBefore.HeadL1.Number, "safeL1", ssBefore.SafeL1.Number, "finalizedL1", ssBefore.FinalizedL1.Number,
+			"unsafeL2", ssBefore.UnsafeL2.Number, "safeL2", ssBefore.SafeL2.Number, "finalizedL2", ssBefore.FinalizedL2.Number)
+		harness.log.Info("Sync status after",
+			"unsafeL1", ssAfter.HeadL1.Number, "safeL1", ssAfter.SafeL1.Number, "finalizedL1", ssAfter.FinalizedL1.Number,
+			"unsafeL2", ssAfter.UnsafeL2.Number, "safeL2", ssAfter.SafeL2.Number, "finalizedL2", ssAfter.FinalizedL2.Number)
 	}
 }
```

samlaf committed:
fix(damgr): ethda failover finalization stall bug

Weiwei from Polymer found this bug. He proposed a solution. This is an alternative solution which seems simpler, but not 100% of its soundness.

samlaf committed:
test(altda): add a test to make sure altda node keeps finalizing even after failover to ethda

Currently it does not, as shown by the test TestAltDA_FinalizationAfterEthDAFailover failing

samlaf commented on op-e2e/actions/altda/altda_test.go:
```diff
@@ -607,3 +643,42 @@ func TestAltDA_Finalization(gt *testing.T) {
+	for i := 0; i < 2; i++ {
+		ssBefore := harness.sequencer.SyncStatus()
+		harness.ActNewL2TxFinalized(t)
+		if i == 0 {
+			// TODO: figure out why this is needed
+			// I think it's because the L1 driven finalizedHead is set to L1FinalizedHead-ChallengeWindow (see damgr.go updateFinalizedFromL1),
+			// so it trails behind by an extra challenge_window when we switch over to ethDA.
+			harness.ActNewL2TxFinalized(t)
+		}
```
TODO: this felt weird to me, but maybe its expected and fine?

alfredo-stonk commented on op-alt-da/damgr.go:
```diff
@@ -117,8 +117,16 @@ func (d *DA) OnFinalizedHeadSignal(f HeadSignalFn) {
+	// If a commitment was pruned, set the finalized head to that commitment's inclusion block
+	// When no commitments are left to be pruned (one example is if we have failed over to ethda)
+	// then updateFinalizedFromL1 becomes the main driver of the finalized head.
+	// Note that updateFinalizedFromL1 is only called when len(d.state.commitments) == 0
```
nit: ... only called when `d.state.NoCommitments()` is true

samlaf replied:
[1bf1f68](https://github.com/ethereum-optimism/optimism/pull/12845/commits/1bf1f6853c0a5783749f059a7dcec3b88747b991)

alfredo-stonk commented on op-e2e/actions/altda/altda_test.go:
```diff
@@ -607,3 +643,42 @@ func TestAltDA_Finalization(gt *testing.T) {
 	// given 12s l1 time and 1s l2 time, l2 should be 12 * 3 = 36 blocks finalized
 	require.Equal(t, uint64(36), a.sequencer.SyncStatus().FinalizedL2.Number)
 }
+
+func TestAltDA_FinalizationAfterEthDAFailover(gt *testing.T) {
```
Nice, this test case covers ethDA -> altDA. Can we also have a similar case ethDA -> altDA -> ethDA, simulating a temp altDA failure?

samlaf replied:
Done: [ce0267b](https://github.com/ethereum-optimism/optimism/pull/12845/commits/ce0267b8fd1576680e1ece4aa36541f5116347a3)
I don't fully understand the finalization behavior. Look at the weird `if i==0` cases. Would appreciate help figuring out whether this behavior is really what we want, and if so, how to best explain it in the comment/assert logic.

alfredo-stonk commented on op-alt-da/damgr.go:
```diff
@@ -117,8 +117,16 @@ func (d *DA) OnFinalizedHeadSignal(f HeadSignalFn) {
+	// then updateFinalizedFromL1 becomes the main driver of the finalized head.
+	// Note that updateFinalizedFromL1 is only called when len(d.state.commitments) == 0
+	var zero eth.L1BlockRef
+	if lastPrunedCommIncBlock != zero {
```
nit: use zero value directly, to be consistent with other zero value cases in the repo

samlaf replied:
[34b2df9](https://github.com/ethereum-optimism/optimism/pull/12845/commits/34b2df99e2a44f55727c13c7eef6d5f683d2916a)

alfredo-stonk reviewed (commented):
Thanks for the fix!

samlaf committed:
ci: migrate from circleci to github actions (#20)

* ci: add test-golang workflow

* ci: delete circleci yml file

This was used by op, we are moving to using github actions for our fork

* ci: fix linter by using golangci-lint action

* ci: use golangci-lint v1.61 to match mise.toml requirement

* ci: use mise in ci to install correct versions of dependencies

* ci: remove setup-go step because go already installed by mise in previous step

* ci: build contracts that are needed for op-e2e tests

* ci: use caching for forge artifacts

* ci: remove slow op-program build

don't think we are using it... testing

* ci: make go-lint job run in parallel, and update version to try to fix errors

* ci: add go modules caching to speed up workflow

* ci: add explicit go mod download

samlaf committed:
fix: damgr_test doesn't compile

samlaf committed:
ci: add op-alt-da tests to ci

samlaf committed:
chore: add more logs to damgr and altda_data_source

gastonponti commented on op-e2e/actions/altda/altda_test.go:
```diff
@@ -607,3 +643,64 @@ func TestAltDA_Finalization(gt *testing.T) {
 	require.Equal(t, uint64(36), a.sequencer.SyncStatus().FinalizedL2.Number)
 }
+
+// This test tests ethDA -> altDA -> ethDA finalization behavior, simulating a temp altDA failure.
```
nit:
altDA -> ethDA -> altDa, right?

samlaf replied:
yes lol, thanks. Btw will be closing this, as its out of date and clearly won't be merged. Please use https://github.com/Layr-Labs/optimism/pull/23 going forward

samlaf replied:
Fixed the typo in [11eb3d1](https://github.com/Layr-Labs/optimism/pull/23/commits/11eb3d1d9c017597acc32e67ee8ab6d97017c04a) btw

samlaf committed:
feat: eigenda-holesky kurtosis devnet (#25)

* feat: add eigenda kurtosis devnet

* chore: rename altda -> eigenda devnet

* ci: add workflow to spin up eigenda kurtosis devnet

* ci: fix kurtosis-devnet.yml -- install foundry toolchain

* docs: update readme devnet name altda-devnet -> eigenda-holesky-devnet

* chore: change kt package back to ethpandaops instead of our fork (altda pr was merged)

* fix(eigenda-devnet): use teku since lighthouse is broken with minimal preset right now

* chore: bump da challenge/resolve windows 1->16 in eigenda-holesky.yaml kt-devnet file

* style: swap challenger and proposer ordering in eigenda-holesky.yaml

* ci: pin kurtosis to version 1.4.4

* ci(kurtosis-devnet): use mise to install dependencies with correct versions (kurtosis + foundry)

* ci: delete kt-install gha (we use mise install now)

* docs: explain why we use GenericCommitment and don't deploy da_challenge contract even for failover

* chore: add eigenda-holesky-devnet-clean justfile command to delete eigenda testnet enclave

samlaf committed:
docs(altda_test): fix typo
